collect_app/src/androidTest/java/org/odk/collect/android/formentry/StringWidgetsTest.java

```diff
@@ -0,0 +1,157 @@
+package org.odk.collect.android.formentry;
+
+import android.Manifest;
+
+import androidx.test.espresso.intent.rule.IntentsTestRule;
+import androidx.test.rule.GrantPermissionRule;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+import org.odk.collect.android.activities.FormEntryActivity;
+import org.odk.collect.android.espressoutils.pages.FormEntryPage;
+import org.odk.collect.android.support.CopyFormRule;
+import org.odk.collect.android.support.ResetStateRule;
+import org.odk.collect.android.test.FormLoadingUtils;
+
+public class StringWidgetsTest {
```

Comment:
I like the focus of these tests but personally it feels to me like it would be better to have this testing at a JUnit/Robolectric level rather than here in Espresso. It seems to be that each of these pieces of behaviour could be tested in test for the Widget concerned - the behaviours don't rely on many different components and seem to be the responsbility of the Widget alone. Tests for individual components can be faster which is a definite advantge but it can also make it easier to have a component's core behaviours tested in one place that you can be running repeatedly while making changes. I guess what I'm aiming for is that if I make a change to `XWidget` I can run `XWidgetTest` and have a pretty high chance of catching any problems.

How do you feel about changing it?

Reply:
You are right I converted my tests.